support-models/src/main/scala/com/gu/support/workers/GiftRecipient.scala

```diff
@@ -34,26 +39,11 @@ object GiftRecipient {
     deliveryDate: LocalDate,
   ) extends GiftRecipient
 
-  val circeDiscriminator = new CirceDiscriminator("giftRecipientType")
-
-  private val discriminatorWeekly = "Weekly"
-  private val discriminatorDigitalSubscription = "DigitalSubscription"
-
-  implicit val e1 = circeDiscriminator.add(deriveEncoder[WeeklyGiftRecipient], discriminatorWeekly)
-  implicit val d1 = deriveDecoder[WeeklyGiftRecipient]
-  implicit val e2 = circeDiscriminator.add(deriveEncoder[DigitalSubscriptionGiftRecipient], discriminatorDigitalSubscription)
-  implicit val d2 = deriveDecoder[DigitalSubscriptionGiftRecipient]
-
-  implicit val encoder: Encoder[GiftRecipient] = Encoder.instance {
-    case c: WeeklyGiftRecipient => c.asJson
-    case c: DigitalSubscriptionGiftRecipient => c.asJson
-  }
+  val discriminatedType = new DiscriminatedType[GiftRecipient]("giftRecipientType")
+  implicit val weeklyCodec = discriminatedType.variant[WeeklyGiftRecipient]("Weekly")
+  implicit val dsCodec = discriminatedType.variant[DigitalSubscriptionGiftRecipient]("DigitalSubscription")
+  implicit val codec = discriminatedType.codec(List(weeklyCodec, dsCodec))
```

Comment:
if we standardise the discriminator field name (e.g. `"_type"`) then we could make this simpler
```suggestion
  implicit val weeklyCodec = deriveVariantCodec[WeeklyGiftRecipient]("Weekly")
  implicit val dsCodec = deriveVariantCodec[DigitalSubscriptionGiftRecipient]("DigitalSubscription")
  implicit val codec = deriveTopLevelCodec[GiftRecipient](List(weeklyCodec, dsCodec))
```

Reply:
We could default it to `_type` and leave the option to override if necessary?

Reply:
If you look at this specific commit, you can see the changes resulting from this simplification
https://github.com/guardian/support-frontend/pull/2792/commits/4612be809e1c9e333383b60fddf4988801149384
not sure if it's better or not, it's moved some compiler errors to runtime, but equally, it's only using normal functions and classes rather than nested classes.
Let me know what you think @rupertbates  - since it's the last commit it's easy to decide whether or not to include it.